### ISSUE TITLE
Most common segfault

### ISSUE THREAD

**Issue by seanmakesgames** [discourse/mini_racer#39]
Here's the top of the stack. It happens at 0x18 or 0x30. Line 41 is accessing a variable that has been marshalled from JS (collection_name):
41:   if collection_name.is_a?(MiniRacer::FailedV8Conversion)

```
app/models/database_connection.rb:41: [BUG] Segmentation fault at 0x00000000000018
ruby 2.2.4p230 (2015-12-16 revision 53155) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0127 p:0017 s:0529 e:000526 METHOD  app/models/database_connection.rb:41
c:0126 p:0070 s:0523 e:000522 METHOD  app/models/database_connection.rb:55
c:0125 p:0017 s:0518 e:000517 BLOCK  app/models/database_connection.rb:14 [FINISH]
c:0124 p:---- s:0514 e:000513 CFUNC  :call
c:0123 p:---- s:0512 e:000511 CFUNC  :eval_unsafe
c:0122 p:0015 s:0508 e:000507 BLOCK  bundle/ruby/2.2.0/bundler/gems/mini_racer-bcb3027e8f79/lib/mini_racer.rb:162
c:0121 p:0006 s:0506 e:000505 BLOCK  bundle/ruby/2.2.0/bundler/gems/mini_racer-bcb3027e8f79/lib/mini_racer.rb:63 [FINISH]
c:0120 p:---- s:0504 e:000503 CFUNC  :synchronize
c:0119 p:0009 s:0501 e:000500 METHOD bundle/ruby/2.2.0/bundler/gems/mini_racer-bcb3027e8f79/lib/mini_racer.rb:63
c:0118 p:0009 s:0498 e:000497 METHOD bundle/ruby/2.2.0/bundler/gems/mini_racer-bcb3027e8f79/lib/mini_racer.rb:160
```

Found this article, which looks like a similar issue:
http://blog.packagecloud.io/eng/2014/11/17/debugging-ruby-gem-segfault/

Going to try for a repro in mini_racer using GC.start with a simpler case. Any thoughts in the interim?

**Comment by SamSaffron** [comment]
Can you confirm this happens as well on latest (2.3.1?)

**Comment by seanmakesgames** [comment]
Oh! You are referring to ruby right?

**Comment by seanmakesgames** [comment]
I assume this is true. It's going to take me a bit of time to get environments up to 2.3.1.

**Comment by seanmakesgames** [comment]
Due to multiple factors, I cannot upgrade to 2.3.1 at this time. My current version is 2.2.4 -
I'll attempt to create a repro in mini_racer based off the linked article. If I get it to work, then I'll try for a test run against 2.3.1

**Comment by seanmakesgames** [comment]
Managed to get the load testing setup working and got 31 instances of this in 25 minutes. So, if we do get a 'fix', it'll be pretty easy to test it. :)

**Comment by seanmakesgames** [comment]
Ok. I have repro for the segfault. Can't tell what it is, but it seems running full context many times in a loop can cause it sometimes. Potentially it's the GC freeing something it shouldn't be.
Originally I had the loop running 50 times, and right before I was going to give up and change the code-- I got the repro, so I pumped it up to 5000 -- obv. not the most efficient test case, and not even consistent, but man, it hits a segfault pretty consistently in a similar way.

```
  def test_segfault
    5000.times do
      GC.start

      context = MiniRacer::Context.new(timeout: 5)
      context.attach("echo", proc{|msg| msg.to_sym.to_s})
      assert_raises(MiniRacer::EvalError) do
        context.eval("while(true) echo('foo');")
      end
    end
  end
```
